blocks/lists.ts

```diff
@@ -947,49 +944,48 @@ blocks['lists_split'] = {
   },
   /**
    * Modify this block to have the correct input and output types.
-   * @param {string} newMode Either 'SPLIT' or 'JOIN'.
-   * @private
-   * @this {Block}
+   *
+   * @param newMode Either 'SPLIT' or 'JOIN'.
    */
-  updateType_: function(newMode) {
+  updateType_: function(this: SplitBlock, newMode: string) {
     const mode = this.getFieldValue('MODE');
     if (mode !== newMode) {
-      const inputConnection = this.getInput('INPUT').connection;
-      inputConnection.setShadowDom(null);
-      const inputBlock = inputConnection.targetBlock();
+      const inputConnection = this.getInput('INPUT')!.connection;
+      inputConnection!.setShadowDom(null);
+      const inputBlock = inputConnection!.targetBlock();
       if (inputBlock) {
-        inputConnection.disconnect();
+        inputConnection!.disconnect();
         if (inputBlock.isShadow()) {
-          inputBlock.dispose();
+          inputBlock.dispose(false);
         } else {
           this.bumpNeighbours();
         }
```

Comment:
Not important right now, but I think all of this code is unnecessary. Any shadow should be disposed automatically from `setShadowDom(null)` and any child block should be disconnected automatically from the `setCheck` calls below.

Might be worth filing an issue about.

Reply:
TODO added referencing #6920, containing note referencing your comment.